Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/WebHDFSDeploy.java

```diff
@@ -0,0 +1,167 @@
+import com.microsoft.azure.hdinsight.sdk.storage.webhdfs.WebHdfsParamsBuilder;
+import com.microsoft.azure.hdinsight.spark.jobs.JobUtils;
+import com.microsoft.azure.sqlbigdata.sdk.cluster.SqlBigDataLivyLinkClusterDetail;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import com.microsoft.azuretools.azurecommons.helpers.Nullable;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.client.utils.URLEncodedUtils;
+import org.apache.http.entity.BufferedHttpEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.InputStreamEntity;
+import rx.Observable;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.UnknownServiceException;
+import java.util.List;
+
+public class WebHDFSDeploy implements Deployable, ILogger {
+    @NotNull
+    private IClusterDetail cluster;
+
+    @Nullable
+    private HttpObservable http;
+
+    @NotNull
+    private List<NameValuePair> createDirReqParams;
+
+    @NotNull
+    private List<NameValuePair> uploadReqParams;
+
+    public WebHDFSDeploy(IClusterDetail cluster, HttpObservable http) {
+        this.cluster = cluster;
+        this.uploadReqParams = new WebHdfsParamsBuilder("CREATE")
+                .setOverwrite("true")
+                .setPermission("777")
+                .build();
+
+        this.createDirReqParams = new WebHdfsParamsBuilder("MKDIRS")
+                .setPermission("777")
+                .build();
+
+        this.http = http;
+        if (http != null) {
+            http.setDefaultRequestConfig(RequestConfig.custom().setExpectContinueEnabled(true).build());
+        } else {
+            throw new IllegalArgumentException("Cluster type doesen't support webHDFS storage type.");
+        }
+    }
+
+
+    @Override
+    public URI createUploadDir(@NotNull String rootPath) {
+        return URI.create(rootPath)
+                .resolve(JobUtils.getFormatPathByDate() + "/");
+    }
+
+    @NotNull
+    public Observable<String> deploy(@NotNull File src, @NotNull URI dest) {
+        //three steps to upload via webhdfs
+        // 1.put request to create new dir
+        // 2.put request to get 307 redirect uri from response
+        // 3.put redirect request with file content as setEntity
+
+        HttpPut req = new HttpPut(dest.toString());
+        return http.request(req, null, this.createDirReqParams, null)
+                .doOnNext(
+                        resp -> {
+                            if (resp.getStatusLine().getStatusCode() != 200) {
+                                Observable.error(new UnknownServiceException("Can not create directory to save artifact using webHDFS storage type"));
```

Comment:
I suppose the error throw can be like this:
```suggestion
                                throw new UnknownServiceException("Can not create directory to save artifact using webHDFS storage type");
```